idl/group0_state_machine.idl.hh

```diff
@@ -27,8 +27,12 @@ struct topology_change {
     std::vector<canonical_mutation> mutations;
```

Comment:
> With this command, it's possible to split `topology_change` command into parts
if it doesn't fit max command size.

https://github.com/scylladb/scylladb/pull/13962#discussion_r1208472786

Reply:
But how is it possible to split it atomically? Is atomicity is not needed?

Reply:
> But how is it possible to split it atomically? Is atomicity is not needed?

That's why I asked for updating the commit message and elaborating, @margdoc.

Reply:
@gleb-cloudius to answer your question: `write_mutations` will be used to persist CDC generation data, which doesn't have to be atomic. If we break in the middle it's fine, we'll just have some garbage partition in the table that nobody uses for anything. If we commit all parts then the last command, `topology_change`, will update the topology state which stores the CDC generation UUID which is a key to a partition in that table which at that point contains all the CDC generation data.

Reply:
But them may be we should not guard those writes with state id? Should they be strictly ordered? May be we can write them not over raft at all?

Reply:
Writing them over Raft guarantees that the last `topology_change` command will observe all that data on every node where it's executed (consistency across configuration changes and so on). So I think it's the simplest thing to use Raft for sending those commands.

Regarding state IDs, indeed we might not need them.

Hmm, now that you mentioned this I see a correctness problem... When we're pulling snapshots, we only pull CDC generation data for the last generation UUID that's saved in topology. So if we start sending data for new generation, then another node joins or some slow node requires a snapshot transfer, that node may not get all this data and we're in trouble. @margdoc we need to think how to fix that.

Reply:
> https://github.com/scylladb/scylladb/pull/13962#discussion_r1208472786

Fixed.

> Hmm, now that you mentioned this I see a correctness problem... When we're pulling snapshots, we only pull CDC generation data for the last generation UUID that's saved in topology. So if we start sending data for new generation, then another node joins or some slow node requires a snapshot transfer, that node may not get all this data and we're in trouble. @margdoc we need to think how to fix that.

We decided to send mutations of CDC generation data that ate being broadcasted but aren't entirely applied yet, in the topology snapshot.